### PR TITLE
Sort list of MPC Tensors

### PR DESCRIPTION
**Pull request by kamathhrishi**
## Description
This PR addresses issue #92. For the same, I have created a new function that takes a list of MPCTensors and returns it in ascending/descending order. 

In order to accommodate sorting a new directory called applications was created. 

## How has this been tested?
I have written tests for the functionality I have implemented. 

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [X] My changes are covered by tests

### CODE REVIEW

File: src/sympc/applications/sort.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+        # Two iterators for traversing the two halves
+        i = 0
+        j = 0
+
+        # Iterator for the main list
+        k = 0
+
+        while i < len(left) and j < len(right):
+            if (left[i] < right[j]).reconstruct():
```
Comment (gmuraru) [inline src/sympc/applications/sort.py:48]: Q: Reconstructing here doesn't reveal information about the values? (to the orchestrater at least?)
Shouldn't we use ```*``` as a "selection"/masking mechanism?
If you had a tensor with only 2 elements to find the first one (ascending order).
```
bigger_right = left < right # Still hidden
first_element = (bigger_right - 1) * right + bigger_right * left
```
^ This is not the most efficient, but I think it is sufficient to showcase how we can interchange two elements?
@LaRiffle

Comment (LaRiffle) [reply src/sympc/applications/sort.py:48]: Yes I agree we should not reconstruct any thing, @gmuraru is a good solution to avoid this!

Comment (kamathhrishi) [reply src/sympc/applications/sort.py:48]: I thought the same about it leaking too many values. But, I didn't know any other approach. I will look into the solution you proposed. Thank you.

**Comment by LaRiffle** [comment]
Maybe you could put in the docstring the name of the algorithm you used: maybe it's divide and conquer?

**Comment by kamathhrishi** [comment]
> Maybe you could put in the docstring the name of the algorithm you used: maybe it's divide and conquer?

Sure, Theo.

File: src/sympc/applications/sort.py (hunk 1 above)
Comment (kamathhrishi) [reply src/sympc/applications/sort.py]: I have incorporated what @gmuraru had suggested. But, for now, it involves a Bubblesort implementation.

File: src/sympc/algorithms/algorithms.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            # Swap if the element found is greater
+            # than the next element
+            check = input_list[j] > input_list[j + 1]
+
+            temp = input_list[j]
+            input_list[j] = (1 - check) * input_list[j] + check * input_list[j + 1]
+            input_list[j + 1] = (1 - check) * input_list[j + 1] + check * temp
+
+    if not ascending:
+
```
Comment (gmuraru) [inline src/sympc/algorithms/algorithms.py:56]: NIT: Remove empty line

File: src/sympc/algorithms/algorithms.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
+"""Applications that make use of primitive MPC operations."""
+
+# stdlib
+from typing import List
+
+# third party
```
Comment (gmuraru) [inline src/sympc/algorithms/algorithms.py:6]: MAYBE: Remove 2nd comment

File: tests/sympc/algorithms/sort_test.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# third party
+import pytest
+import torch
+
+from sympc.algorithms.algorithms import sort
+from sympc.session import Session
+from sympc.session import SessionManager
+from sympc.tensor import MPCTensor
+
+
```
Comment (gmuraru) [inline tests/sympc/algorithms/sort_test.py:10]: MAYBE: Add `ascending` as a `pytest.parametrize`

**Review by gmuraru — CHANGES_REQUESTED** [review]
Check comments

**Comment by kamathhrishi** [comment]
> Check comments

@gmuraru Made the required changes :)

File: tests/sympc/algorithms/sort_test.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    sorted_list_1 = []
+    for i in sorted:
+        sorted_list_1.append(i.reconstruct())
+
+    if ascending:
+
+        assert sorted_list_1 == expected_list
+    else:
+
```
Comment (gmuraru) [inline tests/sympc/algorithms/sort_test.py:43]: NITPICK: Remove empty line here

File: tests/sympc/algorithms/sort_test.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        torch.tensor([6.0]),
+        torch.tensor([18.0]),
+    ]
+
+    sorted_list_1 = []
+    for i in sorted:
+        sorted_list_1.append(i.reconstruct())
+
+    if ascending:
+
```
Comment (gmuraru) [inline tests/sympc/algorithms/sort_test.py:40]: And here

File: src/sympc/algorithms/algorithms.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    for i in range(n - 1):
+        # range(n) also work but outer loop will repeat one time more than needed.
+        # Last i elements are already in place
+        for j in range(0, n - i - 1):
+            # traverse the array from 0 to n-i-1
+            # Swap if the element found is greater
+            # than the next element
+            check = input_list[j] > input_list[j + 1]
+            temp = input_list[j]
+            input_list[j] = (1 - check) * input_list[j] + check * input_list[j + 1]
```
Comment (gmuraru) [inline src/sympc/algorithms/algorithms.py:48]: MAYBE: Save `1 - check` to another variable to don't make the same computation 2 times.

**Review by gmuraru — CHANGES_REQUESTED** [review]
See comments

File: src/sympc/algorithms/algorithms.py (hunk 7 above)
Comment (kamathhrishi) [reply src/sympc/algorithms/algorithms.py]: This saved quite a bit of time :) Thanks

**Comment by kamathhrishi** [comment]
> See comments

@gmuraru Made the requested changes.

**Review by gmuraru — APPROVED** [review]
LGTM!
